### PR TITLE
DRMAAV2-23: DRMSystem interface for PBSPro

### PR DESCRIPTION
**Pull request by vinodchitrali**
#### Issue-ID

* **[DRMAAV2-23](https://pbspro.atlassian.net/browse/DRMAAV2-23)**

#### Problem description
*  DRMSystem interface for PBSPro

#### Cause / Analysis
* Briefly describe root cause/analysis of the problem

#### Solution description
* Define DRMSystem and Implement concrete class for PBSPro. This concrete class is the only way to communicate with PBSPro

#### Checklist:

- [x] I have joined the **[pbspro community forum](http://community.pbspro.org/)**.
- [x] My pull request contains a **single, signed** commit. See **[setting up gpg signature](https://pbspro.atlassian.net/wiki/display/DG/Signing+Your+Git+Commits)**.
- [x] My code follows the **[coding style](https://pbspro.atlassian.net/wiki/display/DG/Coding+Standards)** of this project.
- [ ] My change requires project documentation. See **[required documentation checklist](https://pbspro.atlassian.net/wiki/display/DG/Checklist+for+Developing+Features+and+Bug+Fixes)** for details.
- [ ] I have added documentation in the **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)**.
- [ ] I have added new **unit test(s) to my commit**. *(or)*
- [ ] I have added  **manual test(s) to the Jira ticket and explained why unit testing is not appropriate** for this case.
- [ ] All new and existing automated tests have passed. 
- [x] I have attached **test logs to the Jira ticket** as evidence of testing/verification.

__***For further information please visit the [Developer Guide Home](https://pbspro.atlassian.net/wiki/display/DG/Developer+Guide+Home).***__

### CODE REVIEW

File: inc/DRMSystem.h [hunk 1]
```diff
@@ -0,0 +1,509 @@
...
+	 * @throw DrmCommunicationException - Communication errors while
+	 * 										connecting to DRMS
+	 * @throw DeniedByDrmsException - Connection refused by DRMS
+	 * @throw ImplementationSpecificException - Any implementation specific
+	 * 											errors
+	 *
+	 * @return void* - Implementation specific details returned by DRMS
+	 *
+	 */
+	virtual void* connect(void *details_) throw () = 0;
```
Comment (arungrover) [inline inc/DRMSystem.h]: shouldn't there be a connection object part of this class?

Comment (vinodchitrali) [inline inc/DRMSystem.h]: all are reentrant functions.

File: inc/DRMSystem.h [hunk 2]
```diff
@@ -0,0 +1,509 @@
...
+	 *
+	 * @throw DrmCommunicationException - Communication errors while
+	 * 									sending/receiving data from/to DRMS
+	 * @throw ImplementationSpecificException - Any implementation specific
+	 * 											errors
+	 *
+	 * @return - None
+	 *
+	 */
+	virtual void hold(const Connection & connection_,
```
Comment (vinodchitrali) [inline inc/DRMSystem.h]: fixed

File: inc/PBSProSystem.h [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+			const JobInfo& filter_) throw ();
+	/**
+	 * @brief overridden method from DRMSystem
+	 */
+	virtual MachineInfoList getAllMachines(
+			const Connection & connection_) throw ();
+	/**
+	 * @brief overridden method from DRMSystem
+	 */
+	virtual QueueInfoList getAllgetAllQueues(
```
Comment (vinodchitrali) [inline inc/PBSProSystem.h]: fixed

File: inc/PBSProSystem.h [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+			const ReservationTemplate& reservationTemplate_) throw ();
+	/**
+	 * @brief overridden method from DRMSystem
+	 */
+	virtual void remove(const Connection & connection_,
+			const Reservation& reservation_) throw ();
+	/**
+	 * @brief overridden method from DRMSystem
+	 */
+	virtual ReservationList getAllReservations(
```
Comment (vinodchitrali) [inline inc/PBSProSystem.h]: fixed

File: inc/PBSProSystem.h [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+	virtual void reap(const Connection & connection_, const Job& job_) throw ();
+	/**
+	 * @brief overridden method from DRMSystem
+	 */
+	virtual JobState state(const Connection & connection_,
+			const Job& job_) throw ();
+	/**
+	 * @brief overridden method from DRMSystem
+	 */
+	virtual Job* get(const Connection & connection_,
```
Comment (vinodchitrali) [inline inc/PBSProSystem.h]: fixed

File: inc/PBSProSystem.h [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+ * them - whether embedded or bundled with other software - under a commercial
+ * license agreement.
+ *
+ * Use of Altair’s trademarks, including but not limited to "PBS™",
+ * "PBS Professional®", and "PBS Pro™" and Altair’s logos is subject to Altair's
+ * trademark licensing policies.
+ *
+ */
+
+#ifndef INC_PBSPROSYSTEM_H_
```
Comment (vinodchitrali) [inline inc/PBSProSystem.h]: signifies directory name .. .Fixed

File: inc/PBSProSystem.h (hunk 5 above)
Comment (arungrover) [inline inc/PBSProSystem.h]: This still needs to be changed in this declaration and definition of the derived class.

File: inc/PBSProSystem.h (hunk 3 above)
Comment (arungrover) [inline inc/PBSProSystem.h]: changes needed in this derived class as well.

File: inc/DRMSystem.h (hunk 1 above)
Comment (arungrover) [inline inc/DRMSystem.h]: I get that now. I do have a question about why PBSProSystem is made a singleton? would this mean that a multi-threaded application will all be making use of the single instance of that class but operate of different connection objects?
If that is the case what advantage do we get out of making PBSProSystem as singleton? we can as well have the DRMSystem class store the connection object and then have multiple objects operate on them.

Comment (vinodchitrali) [inline inc/DRMSystem.h]: While implement , tot of two ways to implement. 
1) Use static functions .
2) Using singleton pattern. 

Chose the later bcz, 
It gives interface kind flexibility. One can extend the PBSProSystem. In future, some commonly used attribute or one time retried data of PBS can be placed. Then this is right place. There u need single ton. 

Decoupling is one more reason to go for ADT of DRMSystem

File: inc/DRMSystem.h [hunk 7]
```diff
@@ -0,0 +1,543 @@
...
+	 * @param[in] details_ - Implementation specific details required
+	 * 						to disconnect from DRMS
+	 *
+	 * @throw ImplementationSpecificException - Any implementation specific
+	 * 											errors
+	 *
+	 * @return - None
+	 *
+	 */
+	virtual void disconnect(void *details) throw () = 0;
```
Comment (rampranesh) [inline inc/DRMSystem.h]: Hi @vinodchitrali 
    I feel we need to pass a reference of Connection object to connect and disconnect methods.

Thanks,
Rampranesh M

Comment (vinodchitrali) [inline inc/DRMSystem.h]: fixed

File: src/PBSProSystem.cpp [hunk 8]
```diff
@@ -0,0 +1,213 @@
...
+	throw std::exception();
+}
+
+MachineInfoList PBSProSystem::getAllMachines(
+		const Connection& connection_) throw () {
+	//TODO Add Code here
+	throw std::exception();
+}
+
+QueueInfoList PBSProSystem::getAllgetAllQueues(
```
Comment (arungrover) [inline src/PBSProSystem.cpp]: this needs to change

File: inc/DRMSystem.h (hunk 1 above)
Comment (arungrover) [inline inc/DRMSystem.h]: I'm confused by statement - "I_n future, some commonly used attribute or one time retried data of PBS can be placed._". what is the use case here? 
Also, the implementation here and the Class diagram posted in the design proposal does not really match. 
Here is what design proposal says - "ConnectionPool is a set of active connections with PBSPro. In multi-threaded application each thread uses one connection. ConnectionPool Is Member of DRMSystem
"

Comment (vinodchitrali) [inline inc/DRMSystem.h]: Initially i tot about DRMSName and DRMSVersion are two entities. 

Off-late i realized ConnectionPool in DRMSystem is a tight coupling. Also too many get/return with ConnectionPool. So putting  one level above would be better. 

Is this the case for architectural review again ? Then let me get back to architectural review before i move forward with implementation.

**Comment by vinodchitrali** [comment]
@mike0042, @arungrover,
Updated EDD to make ConnectionPool as part of SessionManger. This will decouple the DRMSystem and ConnectionPool.

**Comment by arungrover** [comment]
looks good to me

**Comment by rampranesh** [comment]
Changes looks good to me.

**Comment by vinodchitrali** [comment]
squashed, Ready for merge

**Pull request merged**